Python/ceval.c

```diff
@@ -1924,6 +1925,79 @@ _PyEval_EvalFrameDefault(PyThreadState *tstate, PyFrameObject *f, int throwflag)
             DISPATCH();
         }
 
+        case TARGET(BINARY_SUBSCR_ADAPTIVE): {
+            SpecializedCacheEntry *cache = GET_CACHE();
+            if (cache->adaptive.counter == 0) {
+                PyObject *sub = TOP();
+                PyObject *container = SECOND();
+                next_instr--;
+                if (_Py_Specialize_BinarySubscr(container, sub, next_instr, cache) < 0) {
+                    goto error;
+                }
+                DISPATCH();
+            }
+            else {
+                STAT_INC(BINARY_SUBSCR, deferred);
+                cache->adaptive.counter--;
+                JUMP_TO_INSTRUCTION(BINARY_SUBSCR);
+            }
+        }
+
+        case TARGET(BINARY_SUBSCR_LIST): {
+            PyObject *sub = TOP();
+            PyObject *list = SECOND();
+            DEOPT_IF(!PyLong_CheckExact(sub), BINARY_SUBSCR);
+            DEOPT_IF(!PyList_CheckExact(list), BINARY_SUBSCR);
+            Py_ssize_t index = PyLong_AsSsize_t(sub);
+            DEOPT_IF(index < 0 || index >= PyList_Size(list), BINARY_SUBSCR);
+            STACK_SHRINK(1);
+            PyObject *res = PyList_GetItem(list, index);
+            Py_XINCREF(res);
+            Py_DECREF(list);
+            Py_DECREF(sub);
+            SET_TOP(res);
+            if (res == NULL)
+                goto error;
```

Comment:
If you reorder this a little, you can maybe avoid the ``XINCREF`` which means one less branch. Same for the other variations below. What do you think?
```suggestion
            Py_DECREF(list);
            Py_DECREF(sub);
            SET_TOP(res);
            if (res == NULL)
                goto error;
            PY_INCREF(res);
```

Reply:
I had this at some point, and there were segfaults. I conjectured that the DECREFS were invoking GC and res got into trouble because its refcnt was 0. (I'm not an expert on this either, so I may be wrong :)

Reply:
Oh yikes, I suspected that at first but I still entertained the idea. Thanks for the explanation!

Reply:
Try this:
```C
            STACK_SHRINK(1);
            Py_DECREF(sub);
            PyObject *res = PyList_GetItem(list, index);
            SET_TOP(res);
            Py_DECREF(list);
            if (res == NULL)
                goto error;
            Py_INCREF(res);
```

Reply:
When I do that it's fine on windows but on the mac the build segfaults (I did make regen-all and make clean first).

Reply:
Indeed, freeing the `list` could free `res`. I should think before I type :slightly_smiling_face: 

Since we have already done bounds checks, `res` cannot be `NULL`.
```C
    PyObject *res = PyList_GET_ITEM(list, index);
    assert(res != NULL);
    Py_INCREF(res);
    STACK_SHRINK(1);
    Py_DECREF(sub);
    SET_TOP(res);
    Py_DECREF(list);
    DISPATCH();
}